Replace THPLayout with at::Layout in Python Argument Parser (#34543)

terhuhf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34584 Replace THPLayout with at::Layout in Python Argument Parser (#34543)**

Summary:

Test Plan:
```
python setup.py develop
python test/test_torch.py
```
Output:
```
...
Ran 3834 tests in 198.825s

OK (skipped=180)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20403330](https://our.internmc.facebook.com/intern/diff/D20403330)

terhuhf committed:
Replace THPLayout with at::Layout in Python Argument Parser (#34543)

Summary:

Test Plan:
```
python setup.py develop
python test/test_torch.py
```
Output:
```
...
Ran 3834 tests in 198.825s

OK (skipped=180)
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -162,8 +163,11 @@ struct PythonArgs {
   inline c10::optional<double> toDoubleOptional(int i);
-  inline const THPLayout& layout(int i);
-  inline const THPLayout& layoutWithDefault(int i, const THPLayout& default_layout);
+  // inline const at::Layout& layout(int i);
```
nit: we can delete these, right? :)

terhuhf replied:
Thanks Ed! I also noticed it as well. I'm going to submit the change really quick.

terhuhf commented:
I'm looking at the test failures. Not sure if they are caused by me :/

ezyang commented:
> I'm looking at the test failures. Not sure if they are caused by me :/

Yes, they're caused by this diff. I see some suspicious spots in the diff, comments coming.

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -1359,15 +1359,13 @@ def make_python_binding_args(declaration):
         layout_arg = {
-            'default': 'torch.strided',
-            'dynamic_type': 'Layout',
+            'default': 'None',
```
You should preserve the default here; actually, the reason the old PR wasn't landable was because changing the defaults in places like this caused other parts of the codebase to break. It would be pretty cool to fix this (cc @bhosmer) but it's much bigger than a bootcamp task.

More concretely, what you need to do is rewrite `py_default_layout` to produce the C++ version (rather than the Python version) and initialize it using this code. Look at dtype to see how we do this in the other cases.

ezyang commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -106,8 +106,8 @@ static PyObject * THPVariable_arange(PyObject* self, PyObject* args, PyObject* k
-    "arange(Scalar end, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
-    "arange(Scalar start, Scalar end, Scalar step=1, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
+    "arange(Scalar end, *, Tensor out=None, ScalarType dtype=None, Layout? layout=None, Device device=None, bool pin_memory=False, bool requires_grad=False)",
+    "arange(Scalar start, Scalar end, Scalar step=1, *, Tensor out=None, ScalarType dtype=None, Layout? layout=None, Device device=None, bool pin_memory=False, bool requires_grad=False)",
```
Undo this change

ezyang commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -173,7 +173,7 @@ static PyObject * THPVariable_range(PyObject* self, PyObject* args, PyObject* kw
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
-    "range(Scalar start, Scalar end, Scalar step=1, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool requires_grad=False)",
+    "range(Scalar start, Scalar end, Scalar step=1, *, Tensor out=None, ScalarType dtype=None, Layout? layout=None, Device device=None, bool requires_grad=False)",
```
Same here, undo this change

ezyang commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -241,8 +240,8 @@ static PyObject * THPVariable_randint(PyObject* self_, PyObject* args, PyObject*
-    "randint(int64_t high, IntArrayRef size, *, Generator generator=None, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool requires_grad=False)",
-    "randint(int64_t low, int64_t high, IntArrayRef size, *, Generator generator=None, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool requires_grad=False)",
+    "randint(int64_t high, IntArrayRef size, *, Generator generator=None, Tensor out=None, ScalarType dtype=None, Layout? layout=None, Device device=None, bool requires_grad=False)",
+    "randint(int64_t low, int64_t high, IntArrayRef size, *, Generator generator=None, Tensor out=None, ScalarType dtype=None, Layout? layout=None, Device device=None, bool requires_grad=False)",
```
And here

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -413,11 +413,11 @@ void FunctionParameter::set_default_str(const std::string& str) {
   } else if (type_ == ParameterType::LAYOUT) {
     if (str == "None") {
-      default_layout = nullptr;
+      // do nothing
```
This is a memory hazard, as when `ParameterType::LAYOUT`, `default_layout` may not be initialized. Maybe add an assert here that it is optional, in this case.

ezyang replied:
(There's actually a lot of hazards here, so https://github.com/pytorch/pytorch/pull/33771 was sort of trying to tighten things up, but I also ran out of time on that PR too :)

terhuhf replied:
Fixed with assert allows_none

terhuhf committed:
Update on "Replace THPLayout with at::Layout in Python Argument Parser (#34543)"


Summary:

Test Plan:
```
python setup.py develop
python test/test_torch.py
```
Output:
```
...
Ran 3834 tests in 198.825s

OK (skipped=180)
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terhuhf commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -413,11 +413,15 @@ void FunctionParameter::set_default_str(const std::string& str) {
+      // value.
+      TORCH_INTERNAL_ASSERT_DEBUG_ONLY(default_layout == at::Layout::Strided
+          || default_layout == at::Layout::Sparse
+          || default_layout == at::Layout::Mkldnn);
```
@ezyang  Here I just check if the value of the enum is valid. Not sure if this is ok. And not sure if I should assign default_layout a value, like Strided by default, in this case.

ezyang replied:
No, this doesn't make sense. `default_layout` is uninitialized memory at this point in time, you're going to fail ASAN if you do this. I suggest that you should just assert that the argument is "optional" (in which case we will correctly skip looking at `default_layout` if someone tries to grab it) and maybe, for good luck, audit the code gen to make sure that this invariant is respected.

terhuhf commented on tools/autograd/gen_python_functions.py:
```diff
@@ -1359,13 +1359,13 @@ def make_python_binding_args(declaration):
 
     if is_factory_function or is_like_or_new_function_with_options:
-        py_default_layout = '*torch::getLayout(self.options().backend())' if is_like_or_new_function_with_options else None
+        py_default_layout = '{layout_from_backend(self.options().backend())}' if is_like_or_new_function_with_options else None
```
oops, added a pair of unnecessary `{}`. Removing them.

ezyang commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -121,13 +121,13 @@ static PyObject * THPVariable_arange(PyObject* self, PyObject* args, PyObject* k
           .dtype(scalarType)
           .device(r.device(4))
-          .layout(r.layout(3).layout)
+          .layout(r.layoutOptional(3))
```
The `Layout` argument here is not optional (there is no question mark after it), so this should be a `r.layout(3)` call. The original PR gives bad guidance here, because it was also trying to change the types of the Layout to be Layout?, but that's no longer in scope for this PR.

Basically, this line should look similar to the dtype/device setters you see above. A bunch of other places need to respect in similar ways.

terhuhf committed:
Update on "Replace THPLayout with at::Layout in Python Argument Parser (#34543)"


Summary:

Test Plan:
```
python setup.py develop
python test/test_torch.py
```
Output:
```
...
Ran 3834 tests in 198.825s

OK (skipped=180)
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ezyang commented:
the test failures are not all your fault; try rebasing to master

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -1359,13 +1359,13 @@ def make_python_binding_args(declaration):
         layout_arg = {
             'default': 'torch.strided',
-            'dynamic_type': 'Layout',
+            'dynamic_type': 'c10::optional<Layout>',
```
optional or not here?

terhuhf replied:
@ezyang  quick question, what is the difference between dynamic_type, type, and simple_type?

ezyang replied:
I'm not 100% sure. However, I think it's something like: `type` is the C++ type of the argument, complete with references. `simple_type` has all the references stripped off, and `optional` converted into the postfix question mark. And `dynamic_type` says what the actual runtime type is, e.g., `const Tensor&` might be the type, but `Tensor` is the `dynamic_type`. The best way to find out for sure, though, is to read the code responsible for generating these.

terhuhf commented on tools/autograd/templates/python_torch_functions.cpp:
```diff
@@ -141,13 +141,13 @@ static PyObject * THPVariable_arange(PyObject* self, PyObject* args, PyObject* k
     } else {
       TORCH_CHECK(!r.toBool(7), " `pin_memory` and `out` parameters are incompatible");
-      check_out_type_matches(r.tensor(3), r.scalartype(4), r.isNone(4), r.layout(5), r.isNone(5),
+      check_out_type_matches(r.tensor(3), r.scalartype(4), r.isNone(4), r.layout(5),
```
@ezyang  Does here look alright? I didn't call `r.isNone` on layout. But `check_out_type_matched` takes an optional of Layout, and in the function it check if it's initialized.

ezyang replied:
Yes, this is fine. The other calls should get updated to also call Optional too, but that's a change for another day.

terhuhf committed:
Update on "Replace THPLayout with at::Layout in Python Argument Parser (#34543)"


Summary:

Test Plan:
```
python setup.py develop
python test/test_torch.py
```
Output:
```
...
Ran 3834 tests in 198.825s

OK (skipped=180)
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ezyang reviewed (approved):
assuming CI passes this LGTM
